src/vmm/src/vstate/vcpu/x86_64.rs

```diff
@@ -0,0 +1,463 @@
+                        .map_err(Error::CpuId)?
+                }
+            }
+        }
+
+        self.fd
+            .set_cpuid2(&self.supported_cpuid)
+            .map_err(Error::VcpuSetCpuid)?;
+
+        arch::x86_64::msr::setup_msrs(&self.fd).map_err(Error::MSRSConfiguration)?;
+        arch::x86_64::regs::setup_regs(&self.fd, kernel_start_addr.raw_value() as u64)
+            .map_err(Error::REGSConfiguration)?;
+        arch::x86_64::regs::setup_fpu(&self.fd).map_err(Error::FPUConfiguration)?;
+        arch::x86_64::regs::setup_sregs(guest_mem, &self.fd).map_err(Error::SREGSConfiguration)?;
+        arch::x86_64::interrupts::set_lint(&self.fd).map_err(Error::LocalIntConfiguration)?;
+        Ok(())
+    }
+
+    /// Save the KVM internal state.
+    pub fn save_state(&self) -> Result<VcpuState> {
+        /*
+         * Ordering requirements:
+         *
+         * KVM_GET_MP_STATE calls kvm_apic_accept_events(), which might modify
+         * vCPU/LAPIC state. As such, it must be done before most everything
+         * else, otherwise we cannot restore everything and expect it to work.
+         *
+         * KVM_GET_VCPU_EVENTS/KVM_SET_VCPU_EVENTS is unsafe if other vCPUs are
+         * still running.
+         *
+         * KVM_GET_LAPIC may change state of LAPIC before returning it.
+         *
+         * GET_VCPU_EVENTS should probably be last to save. The code looks as
+         * it might as well be affected by internal state modifications of the
+         * GET ioctls.
+         *
+         * SREGS saves/restores a pending interrupt, similar to what
+         * VCPU_EVENTS also does.
+         *
+         * GET_MSRS requires a pre-populated data structure to do something
+         * meaningful. For SET_MSRS it will then contain good data.
+         */
+
+        // Build the list of MSRs we want to save.
+        let num_msrs = self.msr_list.as_fam_struct_ref().nmsrs as usize;
+        let mut msrs = Msrs::new(num_msrs);
+        {
+            let indices = self.msr_list.as_slice();
+            let msr_entries = msrs.as_mut_slice();
+            assert_eq!(indices.len(), msr_entries.len());
+            for (pos, index) in indices.iter().enumerate() {
+                msr_entries[pos].index = *index;
+            }
+        }
+        let mp_state = self.fd.get_mp_state().map_err(Error::VcpuGetMpState)?;
+        let regs = self.fd.get_regs().map_err(Error::VcpuGetRegs)?;
+        let sregs = self.fd.get_sregs().map_err(Error::VcpuGetSregs)?;
+        let xsave = self.fd.get_xsave().map_err(Error::VcpuGetXsave)?;
+        let xcrs = self.fd.get_xcrs().map_err(Error::VcpuGetXcrs)?;
+        let debug_regs = self.fd.get_debug_regs().map_err(Error::VcpuGetDebugRegs)?;
+        let lapic = self.fd.get_lapic().map_err(Error::VcpuGetLapic)?;
+        let nmsrs = self.fd.get_msrs(&mut msrs).map_err(Error::VcpuGetMsrs)?;
+        assert_eq!(nmsrs, num_msrs);
```

Comment:
nit: unrelated to this PR, but UH-OH! This leftover `assert` here looks dangerous, it should be a `Error::VcpuGetMsrs` instead of an induced crash.

(the assert around msr indices is fine as that validates code implementation logic, but this one depends on the value returned from `kvm` and can actually trigger with no fault in code logic)